File: clients/go-sdk/tsid/tsid.go
package tsid

import (
	"crypto/rand"
	"encoding/binary"
	"strings"
	"sync/atomic"
	"time"
)

// alphabet is the Crockford Base32 alphabet (excludes I, L, O, U).
var alphabet = []byte("0123456789ABCDEFGHJKMNPQRSTVWXYZ")

var counter atomic.Uint32

// Generate returns a typed ID with the entity's 3-char prefix: "clt_0HZXEQ5Y8JY5Z".
func Generate(e EntityType) string {
	return e.Prefix() + "_" + raw()
}

// GenerateWithPrefix returns a typed ID with a custom 3-char prefix.
// Use this for application-specific entity types not in EntityType.
func GenerateWithPrefix(prefix string) string {
	return prefix + "_" + raw()
}

// GenerateUntyped returns the raw 13-char TSID with no prefix.
// Use for non-entity IDs (execution IDs, trace IDs, outbox message IDs).
func GenerateUntyped() string {
	return raw()
}

// raw builds the 13-char Crockford Base32 TSID.
//   - 42 bits: milliseconds since Unix epoch
//   - 10 bits: cryptographically-random
//   - 12 bits: process-local atomic counter (mod 4096)
func raw() string {
	now := uint64(time.Now().UnixMilli()) & 0x3FFFFFFFFFF
	r := uint64(randomU16()) & 0x3FF
	c := uint64(counter.Add(1)) & 0xFFF
	// Layout: 42 bits timestamp | 10 bits random | 12 bits counter.
	// Matches crates/fc-common/src/tsid.rs exactly.
	v := (now << 22) | (r << 12) | c
	return encodeCrockford(v)
}

// encodeCrockford renders a 64-bit value as 13 Crockford Base32 chars
// (no padding, base 32 = 5 bits per char × 13 = 65 bits).
func encodeCrockford(v uint64) string {
	out := make([]byte, 13)
	for i := 12; i >= 0; i-- {
		out[i] = alphabet[v&0x1F]
		v >>= 5
	}
	return string(out)
}

// DecodeCrockford parses a 13-char Crockford Base32 string back to its
// 64-bit value. Returns 0, false on malformed input.
func DecodeCrockford(s string) (uint64, bool) {
	if len(s) != 13 {
		return 0, false
	}
	var v uint64
	for i := 0; i < 13; i++ {
		c := s[i]
		if c >= 'a' && c <= 'z' {
			c -= 'a' - 'A'
		}
		var d uint64
		switch {
		case c >= '0' && c <= '9':
			d = uint64(c - '0')
		case c >= 'A' && c <= 'H':
			d = uint64(c-'A') + 10
		case c >= 'J' && c <= 'K':
			d = uint64(c-'J') + 18
		case c >= 'M' && c <= 'N':
			d = uint64(c-'M') + 20
		case c >= 'P' && c <= 'T':
			d = uint64(c-'P') + 22
		case c >= 'V' && c <= 'Z':
			d = uint64(c-'V') + 27
		default:
			return 0, false
		}
		v = (v << 5) | d
	}
	return v, true
}

// ToLong converts a TSID to its numeric form. Handles both typed
// ("clt_0HZXEQ5Y8JY5Z") and raw ("0HZXEQ5Y8JY5Z") inputs.
func ToLong(s string) (int64, bool) {
	r := s
	if i := strings.Index(s, "_"); i >= 0 && len(s) > 14 {
		r = s[i+1:]
	}
	v, ok := DecodeCrockford(r)
	return int64(v), ok
}

// FromLong renders a numeric TSID as a raw 13-char string (no prefix).
func FromLong(v int64) string { return encodeCrockford(uint64(v)) }

// randomU16 returns 16 bits of crypto-grade randomness.
func randomU16() uint16 {
	var b [2]byte
	if _, err := rand.Read(b[:]); err != nil {
		// crypto/rand should never fail on supported platforms; fall
		// back to time-based mixing if it does.
		now := uint64(time.Now().UnixNano())
		return uint16(now ^ (now >> 16))
	}
	return binary.BigEndian.Uint16(b[:])
}

File: crates/fc-common/src/tsid.rs
//! TSID Generator
//!
//! Generates Time-Sorted IDs as Crockford Base32 strings.
//! Compatible with Java's TsidGenerator for cross-platform ID consistency.
//!
//! Typed IDs follow the format `{prefix}_{tsid}` (e.g., `clt_0HZXEQ5Y8JY5Z`).

use std::sync::atomic::{AtomicU16, Ordering};
use std::time::{SystemTime, UNIX_EPOCH};

/// Crockford Base32 alphabet (excludes I, L, O, U)
const ALPHABET: &[u8] = b"0123456789ABCDEFGHJKMNPQRSTVWXYZ";

static COUNTER: AtomicU16 = AtomicU16::new(0);

/// Well-known entity type prefixes matching the FlowCatalyst platform.
///
/// Use these with [`TsidGenerator::generate`] for platform-compatible typed IDs.
/// For custom entity types, use [`TsidGenerator::generate_with_prefix`].
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum EntityType {
    Client,
    Principal,
    Application,
    ServiceAccount,
    Role,
    Permission,
    OAuthClient,
    AuthCode,
    LoginAttempt,
    ClientAuthConfig,
    AppClientConfig,
    IdpRoleMapping,
    CorsOrigin,
    AnchorDomain,
    IdentityProvider,
    EmailDomainMapping,
    ClientAccessGrant,
    EventType,
    Event,
    EventRead,
    Connection,
    Subscription,
    DispatchPool,
    DispatchJob,
    DispatchJobRead,
    Schema,
    AuditLog,
    PlatformConfig,
    ConfigAccess,
    PasswordResetToken,
    WebauthnCredential,
    ScheduledJob,
    ScheduledJobInstance,
    ScheduledJobInstanceLog,
    ApplicationOpenApiSpec,
    Process,
}

impl EntityType {
    /// Returns the 3-character prefix for this entity type.
    pub fn prefix(&self) -> &'static str {
        match self {
            EntityType::Client => "clt",
            EntityType::Principal => "prn",
            EntityType::Application => "app",
            EntityType::ServiceAccount => "sac",
            EntityType::Role => "rol",
            EntityType::Permission => "prm",
            EntityType::OAuthClient => "oac",
            EntityType::AuthCode => "acd",
            EntityType::LoginAttempt => "lat",
            EntityType::ClientAuthConfig => "cac",
            EntityType::AppClientConfig => "apc",
            EntityType::IdpRoleMapping => "irm",
            EntityType::CorsOrigin => "cor",
            EntityType::AnchorDomain => "anc",
            EntityType::IdentityProvider => "idp",
            EntityType::EmailDomainMapping => "edm",
            EntityType::ClientAccessGrant => "gnt",
            EntityType::EventType => "evt",
            EntityType::Event => "evn",
            EntityType::EventRead => "evr",
            EntityType::Connection => "con",
            EntityType::Subscription => "sub",
            EntityType::DispatchPool => "dpl",
            EntityType::DispatchJob => "djb",
            EntityType::DispatchJobRead => "djr",
            EntityType::Schema => "sch",
            EntityType::AuditLog => "aud",
            EntityType::PlatformConfig => "pcf",
            EntityType::ConfigAccess => "cfa",
            EntityType::PasswordResetToken => "prt",
            EntityType::WebauthnCredential => "pkc",
            EntityType::ScheduledJob => "sjb",
            EntityType::ScheduledJobInstance => "sji",
            EntityType::ScheduledJobInstanceLog => "sjl",
            EntityType::ApplicationOpenApiSpec => "oas",
            EntityType::Process => "prc",
        }
    }
}

/// TSID Generator for creating unique, time-sorted identifiers.
///
/// # Examples
///
/// ```
/// use fc_common::tsid::{TsidGenerator, EntityType};
///
/// // Generate a typed ID for a known entity type
/// let client_id = TsidGenerator::generate(EntityType::Client);
/// assert!(client_id.starts_with("clt_"));
///
/// // Generate with a custom prefix for your own entity types
/// let order_id = TsidGenerator::generate_with_prefix("ord");
/// assert!(order_id.starts_with("ord_"));
///
/// // Generate an untyped ID (e.g., for execution IDs, trace IDs)
/// let trace_id = TsidGenerator::generate_untyped();
/// assert_eq!(trace_id.len(), 13);
/// ```
pub struct TsidGenerator;

impl TsidGenerator {
    /// Generate a raw TSID as a Crockford Base32 string (13 characters).
    fn generate_raw() -> String {
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .expect("Time went backwards")
            .as_millis() as u64;

        let counter = COUNTER.fetch_add(1, Ordering::SeqCst) as u64;
        let random: u64 = rand_u16() as u64 & 0x3FF;

        // Combine: timestamp (42 bits) | random (10 bits) | counter (12 bits)
        let tsid = ((now & 0x3FFFFFFFFFF) << 22) | (random << 12) | (counter & 0xFFF);

        encode_crockford(tsid)
    }

    /// Generate a typed ID with a platform entity prefix: `{prefix}_{tsid}`.
    ///
    /// Example: `generate(EntityType::Client)` → `"clt_0HZXEQ5Y8JY5Z"`
    pub fn generate(entity_type: EntityType) -> String {
        format!("{}_{}", entity_type.prefix(), Self::generate_raw())
    }

    /// Generate a typed ID with a custom prefix: `{prefix}_{tsid}`.
    ///
    /// Use this for application-specific entity types not covered by [`EntityType`].
    ///
    /// Example: `generate_with_prefix("ord")` → `"ord_0HZXEQ5Y8JY5Z"`
    pub fn generate_with_prefix(prefix: &str) -> String {
        format!("{}_{}", prefix, Self::generate_raw())
    }

    /// Generate an untyped ID for non-entity contexts (execution IDs, trace IDs, etc.)
    pub fn generate_untyped() -> String {
        Self::generate_raw()
    }

    /// Convert a TSID string to its numeric representation.
    /// Handles both typed (`clt_0HZXEQ5Y8JY5Z`) and raw (`0HZXEQ5Y8JY5Z`) formats.
    pub fn to_long(tsid_str: &str) -> Option<i64> {
        let raw = if tsid_str.len() > 14 && tsid_str.contains('_') {
            tsid_str.split('_').nth(1)?
        } else {
            tsid_str
        };
        decode_crockford(raw).map(|v| v as i64)
    }

    /// Convert a numeric TSID to its string representation (raw, no prefix).
    pub fn from_long(value: i64) -> String {
        encode_crockford(value as u64)
    }
}

/// Encode a 64-bit value to Crockford Base32 (13 characters).
fn encode_crockford(mut value: u64) -> String {
    let mut result = [b'0'; 13];

    for i in (0..13).rev() {
        result[i] = ALPHABET[(value & 0x1F) as usize];
        value >>= 5;
    }

    String::from_utf8(result.to_vec()).unwrap()
}

/// Decode a Crockford Base32 string to 64-bit value.
fn decode_crockford(s: &str) -> Option<u64> {
    if s.len() != 13 {
        return None;
    }

    let mut result: u64 = 0;
    for c in s.chars() {
        let c = c.to_ascii_uppercase();
        let val = match c {
            '0'..='9' => c as u64 - '0' as u64,
            'A'..='H' => c as u64 - 'A' as u64 + 10,
            'J'..='K' => c as u64 - 'J' as u64 + 18,
            'M'..='N' => c as u64 - 'M' as u64 + 20,
            'P'..='T' => c as u64 - 'P' as u64 + 22,
            'V'..='Z' => c as u64 - 'V' as u64 + 27,
            _ => return None,
        };
        result = (result << 5) | val;
    }

    Some(result)
}

/// Simple random u16 using system time and counter.
fn rand_u16() -> u16 {
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_nanos() as u64;
    let counter = COUNTER.load(Ordering::Relaxed) as u64;
    ((now ^ (counter.wrapping_mul(0x5851F42D4C957F2D))) & 0xFFFF) as u16
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_generate_typed_id() {
        let id = TsidGenerator::generate(EntityType::Client);
        assert_eq!(id.len(), 17);
        assert!(id.starts_with("clt_"));
    }

    #[test]
    fn test_generate_custom_prefix() {
        let id = TsidGenerator::generate_with_prefix("ord");
        assert_eq!(id.len(), 17);
        assert!(id.starts_with("ord_"));
    }

    #[test]
    fn test_generate_untyped_id() {
        let id = TsidGenerator::generate_untyped();
        assert_eq!(id.len(), 13);
    }

    #[test]
    fn test_uniqueness() {
        let mut ids = std::collections::HashSet::new();
        for _ in 0..1000 {
            let id = TsidGenerator::generate(EntityType::Client);
            assert!(ids.insert(id), "Duplicate TSID generated");
        }
    }

    #[test]
    fn test_round_trip_typed() {
        let id = TsidGenerator::generate(EntityType::Client);
        let num = TsidGenerator::to_long(&id).unwrap();
        let back = TsidGenerator::from_long(num);
        assert_eq!(&id[4..], back);
    }

    #[test]
    fn test_round_trip_raw() {
        let id = TsidGenerator::generate_untyped();
        let num = TsidGenerator::to_long(&id).unwrap();
        let back = TsidGenerator::from_long(num);
        assert_eq!(id, back);
    }

    #[test]
    fn test_sortability() {
        let id1 = TsidGenerator::generate(EntityType::Client);
        std::thread::sleep(std::time::Duration::from_millis(1));
        let id2 = TsidGenerator::generate(EntityType::Client);
        assert!(id1 < id2, "TSIDs should be lexicographically sortable");
    }
}
